metals/src/main/scala/scala/meta/internal/metals/watcher/FileWatcher.scala

```diff
@@ -196,9 +205,20 @@ object FileWatcher {
     }
 
     () => {
-      watcher.close()
-      stopWatchingSignal.set(true)
-      thread.interrupt()
+      val hanged: AtomicReference[Thread] = new AtomicReference(null)
+      val closing = Future {
+        hanged.set(Thread.currentThread())
+        watcher.close()
+        stopWatchingSignal.set(true)
+        thread.interrupt()
+        hanged.set(null)
+      }
+      // watcher.close() might hang so we don't want to get stuck here
+      Try(Await.ready(closing, 3.seconds)) match {
+        case Failure(_: TimeoutException) =>
+          Option(hanged.get()).foreach(_.interrupt())
```

Comment:
Probably it's also needed to call `stopWatchingSignal.set(true)`  to stop `metals-watch-callback-thread`

Reply:
I moved `stopWatchingSignal.set(true)` before close which I found to hang

Reply:
oh, now i got it